File: packages/ios/Sources/CrumbCore/CrumbReportEnvelope.swift
import Foundation

package struct CrumbReportRuntime: Equatable, Sendable {
    package let osVersion: String
    package let deviceFamily: String
    package let locale: String
    package let timezone: String

    package init(osVersion: String, deviceFamily: String, locale: String, timezone: String) {
        self.osVersion = osVersion
        self.deviceFamily = deviceFamily
        self.locale = locale
        self.timezone = timezone
    }
}

package enum CrumbScreenshotCaptureState: String, Equatable, Sendable {
    case enabled
    case disabledByConfiguration = "disabled_by_configuration"
    case unavailable
}

package enum CrumbScreenshotMaskingState: String, Equatable, Sendable {
    case applied
    case notApplicable = "not_applicable"
    case failed
}

package struct CrumbArtifactManifest: Equatable, Sendable {
    package let id: String
    package let kind: String
    package let mimeType: String
    package let byteSize: Int
    package let sha256: String
    package let redactionState: String
    package let uploadID: String

    package init(
        id: String,
        kind: String,
        mimeType: String,
        byteSize: Int,
        sha256: String,
        redactionState: String,
        uploadID: String
    ) {
        self.id = id
        self.kind = kind
        self.mimeType = mimeType
        self.byteSize = byteSize
        self.sha256 = sha256
        self.redactionState = redactionState
        self.uploadID = uploadID
    }
}

package struct CrumbReportBuildInput: Equatable, Sendable {
    package let reportID: String
    package let trigger: CrumbInvocation
    package let triggeredAt: Date
    package let submittedAt: Date
    package let runtime: CrumbReportRuntime
    package let category: String
    package let description: String
    package let diagnostics: CrumbDiagnosticsSnapshot
    package let screenshotCapture: CrumbScreenshotCaptureState
    package let screenshotMasking: CrumbScreenshotMaskingState
    package let artifacts: [CrumbArtifactManifest]

    package init(
        reportID: String,
        trigger: CrumbInvocation,
        triggeredAt: Date,
        submittedAt: Date,
        runtime: CrumbReportRuntime,
        category: String,
        description: String,
        diagnostics: CrumbDiagnosticsSnapshot,
        screenshotCapture: CrumbScreenshotCaptureState,
        screenshotMasking: CrumbScreenshotMaskingState,
        artifacts: [CrumbArtifactManifest] = []
    ) {
        self.reportID = reportID
        self.trigger = trigger
        self.triggeredAt = triggeredAt
        self.submittedAt = submittedAt
        self.runtime = runtime
        self.category = category
        self.description = description
        self.diagnostics = diagnostics
        self.screenshotCapture = screenshotCapture
        self.screenshotMasking = screenshotMasking
        self.artifacts = artifacts
    }
}

package struct CrumbSerializedReportEnvelope: Equatable, Sendable {
    package let reportID: String
    package let submittedAt: Date
    package let data: Data

    package var json: String { String(decoding: data, as: UTF8.self) }
}

package enum CrumbReportEnvelopeError: Error, Equatable {
    case invalidConfiguration
    case invalidReportID
    case invalidTimestampOrder
    case invalidRuntime
    case invalidCategory
    case invalidDescription
    case tooManyArtifacts
    case invalidArtifact
}

package enum CrumbReportEnvelopeBuilder {
    package static func makeReportID(uuid: UUID = UUID()) -> String {
        "rpt_" + uuid.uuidString.replacingOccurrences(of: "-", with: "")
    }

    package static func build(
        settings: CrumbReportSettings,
        input: CrumbReportBuildInput
    ) throws -> CrumbSerializedReportEnvelope {
        try validate(settings: settings, input: input)
        let diagnostics = input.diagnostics
        let category = input.category.trimmingCharacters(in: .whitespacesAndNewlines)
        let description = input.description.trimmingCharacters(in: .whitespacesAndNewlines)
        let memory: MemoryDTO? = if diagnostics.residentMemoryBytes != nil
            || diagnostics.physicalFootprintBytes != nil {
            MemoryDTO(
                residentBytes: diagnostics.residentMemoryBytes,
                physicalFootprintBytes: diagnostics.physicalFootprintBytes
            )
        } else {
            nil
        }
        let health = diagnostics.network.healthCheck.map {
            HealthDTO(
                host: $0.host,
                succeeded: $0.succeeded,
                statusCode: $0.statusCode,
                latencyMS: $0.latencyMilliseconds,
                failure: $0.failure
            )
        }
        let logCapture = switch diagnostics.logs.status {
        case .disabled: "disabled_by_configuration"
        case .unavailable: "unavailable"
        case .captured, .empty: "enabled"
        }

        let envelope = EnvelopeDTO(
            schemaVersion: "1.0",
            reportID: input.reportID,
            trigger: input.trigger.rawValue,
            triggeredAt: input.triggeredAt,
            submittedAt: input.submittedAt,
            release: ReleaseDTO(
                platform: "ios",
                appVersion: settings.release.appVersion,
                nativeBuild: settings.release.nativeBuild,
                jsBundleVersion: settings.release.bundleVersion,
                environment: settings.environment
            ),
            sdk: SDKDTO(name: "crumb-ios", version: CrumbSDKVersion.current, integration: "native"),
            runtime: RuntimeDTO(
                os: "ios",
                osVersion: input.runtime.osVersion,
                deviceFamily: input.runtime.deviceFamily,
                locale: input.runtime.locale,
                timezone: input.runtime.timezone
            ),
            userInput: UserInputDTO(
                category: category.lowercased(),
                description: description
            ),
            diagnostics: DiagnosticsDTO(
                capturedAt: diagnostics.capturedAt,
                location: diagnostics.location,
                process: ProcessDTO(name: diagnostics.processName, id: Int(diagnostics.processID)),
                cpuUsagePercent: finite(diagnostics.cpuUsagePercent),
                memory: memory,
                thermalState: diagnostics.thermalState,
                threads: ThreadsDTO(
                    count: diagnostics.threadCount,
                    busiest: diagnostics.busiestThreads.map {
                        ThreadDTO(
                            id: String($0.id),
                            name: $0.name,
                            state: $0.state,
                            cpuUsagePercent: finite($0.cpuUsagePercent)
                        )
                    }
                ),
                gpu: GPUDTO(status: "unavailable_on_demand"),
                network: NetworkDTO(
                    status: diagnostics.network.status,
                    transport: diagnostics.network.transport,
                    cellularGeneration: normalizeGeneration(diagnostics.network.cellularGeneration),
                    isExpensive: diagnostics.network.isExpensive,
                    isConstrained: diagnostics.network.isConstrained,
                    healthCheck: health
                ),
                logs: LogsDTO(
                    status: diagnostics.logs.status.rawValue,
                    sources: diagnostics.logs.sources,
                    entries: diagnostics.logs.entries.map {
                        LogEntryDTO(
                            timestamp: $0.timestamp,
                            level: $0.level.rawValue,
                            source: $0.source,
                            category: $0.category,
                            message: $0.message
                        )
                    },
                    truncated: diagnostics.logs.truncated,
                    droppedEntryCount: diagnostics.logs.droppedEntryCount,
                    failures: diagnostics.logs.failures
                ),
                stackTraces: StackTracesDTO(
                    status: diagnostics.stackTraces.status.rawValue,
                    scope: diagnostics.stackTraces.scope,
                    threads: diagnostics.stackTraces.threads.map {
                        ThreadStackDTO(
                            id: String($0.id),
                            name: $0.name,
                            state: $0.state,
                            frames: $0.frames
                        )
                    },
                    truncated: diagnostics.stackTraces.truncated,
                    unavailableReason: diagnostics.stackTraces.unavailableReason
                )
            ),
            privacy: PrivacyDTO(
                screenshotCapture: input.screenshotCapture.rawValue,
                screenshotMasking: input.screenshotMasking.rawValue,
                diagnosticsCapture: "on_demand",
                logCapture: logCapture
            ),
            artifacts: input.artifacts.map {
                ArtifactDTO(
                    id: $0.id,
                    kind: $0.kind,
                    mimeType: $0.mimeType,
                    byteSize: $0.byteSize,
                    sha256: $0.sha256,
                    redactionState: $0.redactionState,
                    uploadID: $0.uploadID
                )
            }
        )

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.keyEncodingStrategy = .convertToSnakeCase
        encoder.outputFormatting = [.sortedKeys]
        return CrumbSerializedReportEnvelope(
            reportID: input.reportID,
            submittedAt: input.submittedAt,
            data: try encoder.encode(envelope)
        )
    }

    private static func validate(
        settings: CrumbReportSettings,
        input: CrumbReportBuildInput
    ) throws {
        guard hasLength(settings.environment, maximum: 64),
              hasLength(settings.release.appVersion, maximum: 64),
              hasLength(settings.release.nativeBuild, maximum: 64),
              settings.release.bundleVersion.map({ hasLength($0, maximum: 128) }) ?? true else {
            throw CrumbReportEnvelopeError.invalidConfiguration
        }
        guard input.reportID.range(
            of: #"^rpt_[A-Za-z0-9_-]{16,80}$"#,
            options: .regularExpression
        ) != nil else {
            throw CrumbReportEnvelopeError.invalidReportID
        }
        guard input.triggeredAt.timeIntervalSinceReferenceDate.isFinite,
              input.submittedAt.timeIntervalSinceReferenceDate.isFinite,
              input.triggeredAt <= input.submittedAt else {
            throw CrumbReportEnvelopeError.invalidTimestampOrder
        }
        guard hasLength(input.runtime.osVersion, maximum: 64),
              hasLength(input.runtime.deviceFamily, maximum: 128),
              hasLength(input.runtime.locale, maximum: 64),
              hasLength(input.runtime.timezone, maximum: 128) else {
            throw CrumbReportEnvelopeError.invalidRuntime
        }
        let category = input.category.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !category.isEmpty, category.unicodeScalars.count <= 64 else {
            throw CrumbReportEnvelopeError.invalidCategory
        }
        let description = input.description.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !description.isEmpty, description.unicodeScalars.count <= 4_000 else {
            throw CrumbReportEnvelopeError.invalidDescription
        }
        guard input.artifacts.count <= 10 else {
            throw CrumbReportEnvelopeError.tooManyArtifacts
        }
        let artifactIDPattern = try! NSRegularExpression(pattern: "^art_[A-Za-z0-9_-]{12,80}$")
        let uploadIDPattern = try! NSRegularExpression(pattern: "^upl_[A-Za-z0-9_-]{12,80}$")
        let digestPattern = try! NSRegularExpression(pattern: "^[a-f0-9]{64}$")
        let kinds = Set(["screenshot", "annotated_screenshot", "other"])
        let redactionStates = Set(["sanitized", "masked", "not_applicable"])
        for artifact in input.artifacts {
            guard matches(artifact.id, expression: artifactIDPattern),
                  kinds.contains(artifact.kind),
                  hasLength(artifact.mimeType, maximum: 128),
                  artifact.byteSize >= 0,
                  artifact.byteSize <= 26_214_400,
                  matches(artifact.sha256, expression: digestPattern),
                  redactionStates.contains(artifact.redactionState),
                  matches(artifact.uploadID, expression: uploadIDPattern) else {
                throw CrumbReportEnvelopeError.invalidArtifact
            }
        }
    }

    private static func hasLength(_ value: String, maximum: Int) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return !trimmed.isEmpty && value.unicodeScalars.count <= maximum
    }

    private static func matches(_ value: String, expression: NSRegularExpression) -> Bool {
        let range = NSRange(value.startIndex..., in: value)
        return expression.firstMatch(in: value, range: range)?.range == range
    }

    private static func finite(_ value: Double?) -> Double? {
        value?.isFinite == true ? value : nil
    }

    private static func normalizeGeneration(_ value: String?) -> String? {
        switch value?.lowercased() {
        case "2g": "2g"
        case "3g": "3g"
        case "4g/lte", "4g_lte": "4g_lte"
        case "5g": "5g"
        case .some: "unknown"
        case nil: nil
        }
    }
}

private struct EnvelopeDTO: Encodable {
    let schemaVersion: String
    let reportID: String
    let trigger: String
    let triggeredAt: Date
    let submittedAt: Date
    let release: ReleaseDTO
    let sdk: SDKDTO
    let runtime: RuntimeDTO
    let userInput: UserInputDTO
    let diagnostics: DiagnosticsDTO
    let privacy: PrivacyDTO
    let artifacts: [ArtifactDTO]
}

private struct ReleaseDTO: Encodable {
    let platform: String
    let appVersion: String
    let nativeBuild: String
    let jsBundleVersion: String?
    let environment: String
}

private struct SDKDTO: Encodable {
    let name: String
    let version: String
    let integration: String
}

private struct RuntimeDTO: Encodable {
    let os: String
    let osVersion: String
    let deviceFamily: String
    let locale: String
    let timezone: String
}

private struct UserInputDTO: Encodable {
    let category: String
    let description: String
}

private struct DiagnosticsDTO: Encodable {
    let capturedAt: Date
    let location: String
    let process: ProcessDTO
    let cpuUsagePercent: Double?
    let memory: MemoryDTO?
    let thermalState: String
    let threads: ThreadsDTO
    let gpu: GPUDTO
    let network: NetworkDTO
    let logs: LogsDTO
    let stackTraces: StackTracesDTO
}

private struct ProcessDTO: Encodable { let name: String; let id: Int }

private struct MemoryDTO: Encodable {
    let residentBytes: UInt64?
    let physicalFootprintBytes: UInt64?
}

private struct ThreadsDTO: Encodable {
    let count: Int
    let busiest: [ThreadDTO]
}

private struct ThreadDTO: Encodable {
    let id: String
    let name: String
    let state: String
    let cpuUsagePercent: Double?
}

private struct GPUDTO: Encodable { let status: String }

private struct NetworkDTO: Encodable {
    let status: String
    let transport: String
    let cellularGeneration: String?
    let isExpensive: Bool
    let isConstrained: Bool
    let healthCheck: HealthDTO?
}

private struct HealthDTO: Encodable {
    let host: String
    let succeeded: Bool
    let statusCode: Int?
    let latencyMS: Int
    let failure: String?

    private enum CodingKeys: String, CodingKey {
        case host, succeeded, failure
        case statusCode = "status_code"
        case latencyMS = "latency_ms"
    }
}

private struct LogsDTO: Encodable {
    let status: String
    let sources: [String]
    let entries: [LogEntryDTO]
    let truncated: Bool
    let droppedEntryCount: Int
    let failures: [String]
}

private struct LogEntryDTO: Encodable {
    let timestamp: Date
    let level: String
    let source: String
    let category: String
    let message: String
}

private struct StackTracesDTO: Encodable {
    let status: String
    let scope: String
    let threads: [ThreadStackDTO]
    let truncated: Bool
    let unavailableReason: String?
}

private struct ThreadStackDTO: Encodable {
    let id: String
    let name: String
    let state: String
    let frames: [String]
}

private struct PrivacyDTO: Encodable {
    let screenshotCapture: String
    let screenshotMasking: String
    let diagnosticsCapture: String
    let logCapture: String
}

private struct ArtifactDTO: Encodable {
    let id: String
    let kind: String
    let mimeType: String
    let byteSize: Int
    let sha256: String
    let redactionState: String
    let uploadID: String
}

File: packages/android/crumb-core/src/main/kotlin/dev/crumb/core/CrumbReportEnvelope.kt
package dev.crumb.core

import java.time.Instant
import java.util.UUID

data class CrumbReportRuntime(
    val osVersion: String,
    val deviceFamily: String,
    val locale: String,
    val timezone: String,
)

enum class CrumbScreenshotCaptureState(val wireValue: String) {
    ENABLED("enabled"),
    DISABLED_BY_CONFIGURATION("disabled_by_configuration"),
    DISABLED_BY_POLICY("disabled_by_policy"),
    UNAVAILABLE("unavailable"),
}

enum class CrumbScreenshotMaskingState(val wireValue: String) {
    APPLIED("applied"),
    NOT_APPLICABLE("not_applicable"),
    FAILED("failed"),
}

data class CrumbArtifactManifest(
    val id: String,
    val kind: String,
    val mimeType: String,
    val byteSize: Long,
    val sha256: String,
    val redactionState: String,
    val uploadId: String,
)

data class CrumbReportBuildInput(
    val reportId: String,
    val trigger: CrumbInvocation,
    val triggeredAtMillis: Long,
    val submittedAtMillis: Long,
    val runtime: CrumbReportRuntime,
    val category: String,
    val description: String,
    val diagnostics: CrumbDiagnosticsSnapshot,
    val screenshotCapture: CrumbScreenshotCaptureState,
    val screenshotMasking: CrumbScreenshotMaskingState,
    val artifacts: List<CrumbArtifactManifest> = emptyList(),
    val customContext: Map<String, String> = emptyMap(),
    val policyStatus: CrumbPolicyStatus = CrumbPolicyStatus.NOT_CONFIGURED,
    val workspacePolicyVersion: Int? = null,
)

data class CrumbSerializedReportEnvelope(
    val reportId: String,
    val submittedAtMillis: Long,
    val json: String,
)

sealed class CrumbReportEnvelopeException(message: String) : IllegalArgumentException(message) {
    class InvalidConfiguration : CrumbReportEnvelopeException("configuration metadata is invalid")
    class InvalidReportId : CrumbReportEnvelopeException("reportId is invalid")
    class InvalidTimestampOrder : CrumbReportEnvelopeException("submittedAt must not precede triggeredAt")
    class InvalidRuntime : CrumbReportEnvelopeException("runtime metadata is invalid")
    class InvalidCategory : CrumbReportEnvelopeException("category must contain 1-64 characters")
    class InvalidDescription : CrumbReportEnvelopeException("description must contain 1-4000 characters")
    class TooManyArtifacts : CrumbReportEnvelopeException("a report may contain at most 10 artifacts")
    class InvalidArtifact : CrumbReportEnvelopeException("artifact manifest is invalid")
}

internal object CrumbReportEnvelopeBuilder {
    fun makeReportId(uuid: UUID = UUID.randomUUID()): String =
        "rpt_${uuid.toString().replace("-", "")}"

    fun build(
        settings: CrumbReportSettings,
        input: CrumbReportBuildInput,
    ): CrumbSerializedReportEnvelope {
        validate(settings, input)
        val diagnostics = effectiveDiagnostics(input.diagnostics, settings)
        val category = input.category.trim()
        val description = input.description.trim()
        val screenshotEnabled = settings.capture.screenshot &&
            CrumbEvidenceCategory.SCREENSHOT in settings.evidence
        val screenshotCapture = when {
            !settings.capture.screenshot -> CrumbScreenshotCaptureState.DISABLED_BY_CONFIGURATION
            CrumbEvidenceCategory.SCREENSHOT !in settings.evidence ->
                CrumbScreenshotCaptureState.DISABLED_BY_POLICY
            else -> input.screenshotCapture
        }
        val screenshotMasking = if (screenshotEnabled) {
            input.screenshotMasking
        } else {
            CrumbScreenshotMaskingState.NOT_APPLICABLE
        }
        val artifacts = if (screenshotEnabled) input.artifacts else emptyList()
        val memory = if (
            diagnostics.residentMemoryBytes != null || diagnostics.physicalFootprintBytes != null
        ) {
            obj(
                "resident_bytes" to diagnostics.residentMemoryBytes,
                "physical_footprint_bytes" to diagnostics.physicalFootprintBytes,
            )
        } else {
            null
        }
        val health = diagnostics.network.healthCheck?.let {
            obj(
                "host" to it.host,
                "succeeded" to it.succeeded,
                "status_code" to it.statusCode,
                "latency_ms" to it.latencyMilliseconds,
                "failure" to it.failure,
            )
        }
        val logCapture = when (diagnostics.logs.status) {
            CrumbLogCaptureStatus.DISABLED -> "disabled_by_configuration"
            CrumbLogCaptureStatus.DISABLED_BY_POLICY -> "disabled_by_policy"
            CrumbLogCaptureStatus.UNAVAILABLE -> "unavailable"
            CrumbLogCaptureStatus.CAPTURED,
            CrumbLogCaptureStatus.EMPTY,
            -> "enabled"
        }

        val envelope = obj(
            "schema_version" to "1.0",
            "report_id" to input.reportId,
            "trigger" to input.trigger.name.lowercase(),
            "triggered_at" to instant(input.triggeredAtMillis),
            "submitted_at" to instant(input.submittedAtMillis),
            "release" to obj(
                "platform" to "android",
                "app_version" to settings.release.appVersion,
                "native_build" to settings.release.nativeBuild,
                "js_bundle_version" to settings.release.bundleVersion,
                "environment" to settings.environment,
            ),
            "application" to settings.application.name?.let { obj("name" to it) },
            "sdk" to obj(
                "name" to "crumb-android",
                "version" to CrumbSDKVersion.CURRENT,
                "integration" to "native",
            ),
            "runtime" to obj(
                "os" to "android",
                "os_version" to input.runtime.osVersion,
                "device_family" to input.runtime.deviceFamily,
                "locale" to input.runtime.locale,
                "timezone" to input.runtime.timezone,
            ),
            "user_input" to obj(
                "category" to category.lowercase(),
                "description" to description,
            ),
            "diagnostics" to obj(
                "captured_at" to instant(diagnostics.capturedAtMillis),
                "location" to diagnostics.location,
                "process" to obj(
                    "name" to diagnostics.processName,
                    "id" to diagnostics.processId,
                ),
                "cpu_usage_percent" to diagnostics.cpuUsagePercent.finiteOrNull(),
                "memory" to memory,
                "thermal_state" to diagnostics.thermalState,
                "threads" to obj(
                    "count" to diagnostics.threadCount,
                    "busiest" to diagnostics.busiestThreads.map {
                        obj(
                            "id" to it.id.toString(),
                            "name" to it.name,
                            "state" to it.state,
                            "cpu_usage_percent" to it.cpuUsagePercent.finiteOrNull(),
                        )
                    },
                ),
                "gpu" to obj("status" to "unavailable_on_demand"),
                "network" to obj(
                    "status" to diagnostics.network.status,
                    "transport" to diagnostics.network.transport,
                    "cellular_generation" to normalizeGeneration(diagnostics.network.cellularGeneration),
                    "is_expensive" to diagnostics.network.isExpensive,
                    "is_constrained" to diagnostics.network.isConstrained,
                    "health_check" to health,
                ),
                "logs" to obj(
                    "status" to diagnostics.logs.status.name.lowercase(),
                    "sources" to diagnostics.logs.sources,
                    "entries" to diagnostics.logs.entries.map {
                        obj(
                            "timestamp" to instant(it.timestampMillis),
                            "level" to it.level.name.lowercase(),
                            "source" to it.source,
                            "category" to it.category,
                            "message" to it.message,
                        )
                    },
                    "truncated" to diagnostics.logs.truncated,
                    "dropped_entry_count" to diagnostics.logs.droppedEntryCount,
                    "failures" to diagnostics.logs.failures,
                ),
                "stack_traces" to obj(
                    "status" to diagnostics.stackTraces.status.name.lowercase(),
                    "scope" to diagnostics.stackTraces.scope,
                    "threads" to diagnostics.stackTraces.threads.map {
                        obj(
                            "id" to it.id.toString(),
                            "name" to it.name,
                            "state" to it.state,
                            "frames" to it.frames,
                        )
                    },
                    "truncated" to diagnostics.stackTraces.truncated,
                    "unavailable_reason" to diagnostics.stackTraces.unavailableReason,
                ),
            ),
            "privacy" to obj(
                "screenshot_capture" to screenshotCapture.wireValue,
                "screenshot_masking" to screenshotMasking.wireValue,
                "diagnostics_capture" to "on_demand",
                "log_capture" to logCapture,
                "policy_status" to settings.policyStatus.wireValue,
                "policy_version" to settings.workspacePolicyVersion,
            ),
            "custom_context" to settings.customContext.takeIf { it.isNotEmpty() },
            "artifacts" to artifacts.map {
                obj(
                    "id" to it.id,
                    "kind" to it.kind,
                    "mime_type" to it.mimeType,
                    "byte_size" to it.byteSize,
                    "sha256" to it.sha256,
                    "redaction_state" to it.redactionState,
                    "upload_id" to it.uploadId,
                )
            },
        )

        return CrumbSerializedReportEnvelope(
            reportId = input.reportId,
            submittedAtMillis = input.submittedAtMillis,
            json = JsonValueEncoder.encode(envelope),
        )
    }

    private fun validate(settings: CrumbReportSettings, input: CrumbReportBuildInput) {
        if (
            !settings.environment.hasLength(64) ||
            !settings.release.appVersion.hasLength(64) ||
            !settings.release.nativeBuild.hasLength(64) ||
            settings.release.bundleVersion?.hasLength(128) == false
        ) {
            throw CrumbReportEnvelopeException.InvalidConfiguration()
        }
        if (!input.reportId.matches(Regex("^rpt_[A-Za-z0-9_-]{16,80}$"))) {
            throw CrumbReportEnvelopeException.InvalidReportId()
        }
        if (input.triggeredAtMillis > input.submittedAtMillis) {
            throw CrumbReportEnvelopeException.InvalidTimestampOrder()
        }
        if (
            !input.runtime.osVersion.hasLength(64) ||
            !input.runtime.deviceFamily.hasLength(128) ||
            !input.runtime.locale.hasLength(64) ||
            !input.runtime.timezone.hasLength(128)
        ) {
            throw CrumbReportEnvelopeException.InvalidRuntime()
        }
        if (input.category.trim().length !in 1..64) {
            throw CrumbReportEnvelopeException.InvalidCategory()
        }
        if (input.description.trim().length !in 1..4_000) {
            throw CrumbReportEnvelopeException.InvalidDescription()
        }
        if (input.artifacts.size > 10) {
            throw CrumbReportEnvelopeException.TooManyArtifacts()
        }
        if (input.artifacts.any {
                !it.id.matches(Regex("^art_[A-Za-z0-9_-]{12,80}$")) ||
                    it.kind !in setOf("screenshot", "annotated_screenshot", "other") ||
                    !it.mimeType.hasLength(128) ||
                    it.byteSize !in 0..26_214_400 ||
                    !it.sha256.matches(Regex("^[a-f0-9]{64}$")) ||
                    it.redactionState !in setOf("sanitized", "masked", "not_applicable") ||
                    !it.uploadId.matches(Regex("^upl_[A-Za-z0-9_-]{12,80}$"))
            }
        ) {
            throw CrumbReportEnvelopeException.InvalidArtifact()
        }
    }

    private fun effectiveDiagnostics(
        input: CrumbDiagnosticsSnapshot,
        settings: CrumbReportSettings,
    ): CrumbDiagnosticsSnapshot {
        val performanceEnabled = CrumbEvidenceCategory.PERFORMANCE in settings.evidence
        val networkEnabled = CrumbEvidenceCategory.NETWORK in settings.evidence
        val healthCheckEnabled = networkEnabled &&
            CrumbEvidenceCategory.HEALTH_CHECK in settings.evidence
        val logsEnabled = CrumbEvidenceCategory.LOGS in settings.evidence
        val stacksEnabled = CrumbEvidenceCategory.THREAD_STACKS in settings.evidence
        return CrumbDiagnosticsSnapshot(
            capturedAtMillis = input.capturedAtMillis,
            location = input.location,
            processName = input.processName,
            processId = input.processId,
            cpuUsagePercent = if (performanceEnabled) input.cpuUsagePercent else null,
            residentMemoryBytes = if (performanceEnabled) input.residentMemoryBytes else null,
            physicalFootprintBytes = if (performanceEnabled) input.physicalFootprintBytes else null,
            thermalState = if (performanceEnabled) input.thermalState else "unavailable",
            threadCount = if (performanceEnabled) input.threadCount else 0,
            busiestThreads = if (performanceEnabled) input.busiestThreads else emptyList(),
            gpuStatus = if (performanceEnabled) input.gpuStatus else "unavailable_by_policy",
            network = if (networkEnabled) {
                input.network.copy(
                    healthCheck = if (healthCheckEnabled) input.network.healthCheck else null,
                )
            } else {
                CrumbNetworkDiagnostic(
                    status = "unknown",
                    transport = "unknown",
                    cellularGeneration = null,
                    isExpensive = false,
                    isConstrained = false,
                    healthCheck = null,
                )
            },
            logs = if (logsEnabled) {
                input.logs
            } else {
                CrumbLogDiagnostic(
                    status = if (settings.diagnostics.logs.enabled) {
                        CrumbLogCaptureStatus.DISABLED_BY_POLICY
                    } else {
                        CrumbLogCaptureStatus.DISABLED
                    },
                    sources = emptyList(),
                    entries = emptyList(),
                    truncated = false,
                    droppedEntryCount = 0,
                    failures = emptyList(),
                )
            },
            stackTraces = if (stacksEnabled) {
                input.stackTraces
            } else {
                CrumbStackTraceDiagnostic(
                    status = CrumbStackTraceCaptureStatus.UNAVAILABLE,
                    scope = "none",
                    threads = emptyList(),
                    truncated = false,
                    unavailableReason = "disabled_by_policy",
                )
            },
        )
    }

    private fun String.hasLength(maximum: Int): Boolean = isNotBlank() && length <= maximum

    private fun instant(milliseconds: Long): String = Instant.ofEpochMilli(milliseconds).toString()

    private fun normalizeGeneration(value: String?): String? = when (value?.lowercase()) {
        "2g" -> "2g"
        "3g" -> "3g"
        "4g/lte", "4g_lte" -> "4g_lte"
        "5g" -> "5g"
        null -> null
        else -> "unknown"
    }

    private fun Double?.finiteOrNull(): Double? = this?.takeIf(Double::isFinite)

    private fun obj(vararg values: Pair<String, Any?>): Map<String, Any?> = buildMap {
        values.forEach { (key, value) -> if (value != null) put(key, value) }
    }
}

private object JsonValueEncoder {
    fun encode(value: Any?): String = buildString { appendValue(value) }

    private fun StringBuilder.appendValue(value: Any?) {
        when (value) {
            null -> append("null")
            is String -> appendString(value)
            is Boolean -> append(if (value) "true" else "false")
            is Byte, is Short, is Int, is Long -> append(value.toString())
            is Float -> append(if (value.isFinite()) value.toString() else "null")
            is Double -> append(if (value.isFinite()) value.toString() else "null")
            is Map<*, *> -> {
                append('{')
                value.entries.forEachIndexed { index, entry ->
                    if (index > 0) append(',')
                    appendString(entry.key as String)
                    append(':')
                    appendValue(entry.value)
                }
                append('}')
            }
            is Iterable<*> -> {
                append('[')
                value.forEachIndexed { index, item ->
                    if (index > 0) append(',')
                    appendValue(item)
                }
                append(']')
            }
            else -> error("Unsupported JSON value: ${value.javaClass.name}")
        }
    }

    private fun StringBuilder.appendString(value: String) {
        append('"')
        value.forEach { character ->
            when (character) {
                '"' -> append("\\\"")
                '\\' -> append("\\\\")
                '\b' -> append("\\b")
                '\u000C' -> append("\\f")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> if (character.code < 0x20) {
                    append("\\u")
                    append(character.code.toString(16).padStart(4, '0'))
                } else {
                    append(character)
                }
            }
        }
        append('"')
    }
}
